File: app/src/main/java/me/zhanghai/android/files/ui/CrossfadeSubtitleToolbar.java
/*
 * Copyright (c) 2018 Hai Zhang <dreaming.in.code.zh@gmail.com>
 * All Rights Reserved.
 */

package me.zhanghai.android.files.ui;

import android.animation.Animator;
import android.animation.AnimatorListenerAdapter;
import android.animation.ObjectAnimator;
import android.animation.ValueAnimator;
import android.content.Context;
import android.text.TextUtils;
import android.util.AttributeSet;
import android.widget.TextView;

import java.lang.reflect.Field;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.appcompat.widget.Toolbar;
import androidx.interpolator.view.animation.FastOutSlowInInterpolator;
import me.zhanghai.android.files.util.ViewUtils;

public class CrossfadeSubtitleToolbar extends Toolbar {

    @NonNull
    private ObjectAnimator mSubtitleAnimator;

    @Nullable
    private CharSequence mNextSubtitle;

    public CrossfadeSubtitleToolbar(@NonNull Context context) {
        super(context);

        init();
    }

    public CrossfadeSubtitleToolbar(@NonNull Context context, @Nullable AttributeSet attrs) {
        super(context, attrs);

        init();
    }

    public CrossfadeSubtitleToolbar(@NonNull Context context, @Nullable AttributeSet attrs,
                                    int defStyleAttr) {
        super(context, attrs, defStyleAttr);

        init();
    }

    private void init() {
        mSubtitleAnimator = ObjectAnimator.ofFloat(null, ALPHA, 1, 0, 1)
                .setDuration(2 * ViewUtils.getShortAnimTime(this));
        mSubtitleAnimator.setInterpolator(new FastOutSlowInInterpolator());
        AnimatorListener listener = new AnimatorListener();
        mSubtitleAnimator.addUpdateListener(listener);
        mSubtitleAnimator.addListener(listener);
    }

    @Nullable
    @Override
    public CharSequence getSubtitle() {
        if (mNextSubtitle != null) {
            return mNextSubtitle;
        }
        return super.getSubtitle();
    }

    @Override
    public void setSubtitle(@Nullable CharSequence subtitle) {

        if (TextUtils.equals(getSubtitle(), subtitle)) {
            return;
        }

        mNextSubtitle = subtitle;
        ensureSubtitleAnimatorTarget();
        if (mSubtitleAnimator.getTarget() == null) {
            // Subtitle text view not available (yet), just delegate to super.
            super.setSubtitle(subtitle);
            return;
        }
        if (!mSubtitleAnimator.isRunning()) {
            mSubtitleAnimator.start();
        }
    }

    private void ensureSubtitleAnimatorTarget() {
        if (mSubtitleAnimator.getTarget() != null) {
            return;
        }
        TextView subtitleTextView;
        //noinspection TryWithIdenticalCatches
        try {
            Field subtitleTextViewField = Toolbar.class.getDeclaredField("mSubtitleTextView");
            subtitleTextViewField.setAccessible(true);
            subtitleTextView = (TextView) subtitleTextViewField.get(this);
        } catch (NoSuchFieldException e) {
            e.printStackTrace();
            return;
        } catch (IllegalAccessException e) {
            e.printStackTrace();
            return;
        }
        if (subtitleTextView == null) {
            return;
        }
        // HACK: Prevent setText() from calling requestLayout() during animation which triggers
        // re-layout of the entire view hierarchy and breaks the ripple of BreadcrumbLayout.
        ViewUtils.setWidth(subtitleTextView, LayoutParams.MATCH_PARENT);
        mSubtitleAnimator.setTarget(subtitleTextView);
    }

    private class AnimatorListener extends AnimatorListenerAdapter
            implements ValueAnimator.AnimatorUpdateListener {

        private boolean mTextUpdated;

        @Override
        public void onAnimationUpdate(@NonNull ValueAnimator animator) {
            if (animator.getAnimatedFraction() < 0.5) {
                mTextUpdated = false;
            } else {
                ensureTextUpdated();
            }
        }

        @Override
        public void onAnimationEnd(@NonNull Animator animator) {
            ensureTextUpdated();
            if (mNextSubtitle != null) {
                mTextUpdated = false;
                animator.start();
            }
        }

        private void ensureTextUpdated() {
            if (!mTextUpdated) {
                if (mNextSubtitle != null) {
                    CrossfadeSubtitleToolbar.super.setSubtitle(mNextSubtitle);
                    mNextSubtitle = null;
                }
                mTextUpdated = true;
            }
        }
    }
}

File: app/src/main/java/me/zhanghai/android/files/ui/CrossfadeSubtitleToolbar.kt
/*
 * Copyright (c) 2018 Hai Zhang <dreaming.in.code.zh@gmail.com>
 * All Rights Reserved.
 */

package me.zhanghai.android.files.ui

import android.animation.Animator
import android.animation.AnimatorListenerAdapter
import android.animation.ObjectAnimator
import android.animation.ValueAnimator
import android.animation.ValueAnimator.AnimatorUpdateListener
import android.content.Context
import android.util.AttributeSet
import android.view.View
import android.widget.TextView
import androidx.annotation.AttrRes
import androidx.appcompat.widget.Toolbar
import androidx.interpolator.view.animation.FastOutSlowInInterpolator
import me.zhanghai.android.files.util.shortAnimTime

class CrossfadeSubtitleToolbar : Toolbar {
    private val subtitleAnimator = ObjectAnimator.ofFloat(null, View.ALPHA, 1f, 0f, 1f).apply {
        duration = (2 * context.shortAnimTime).toLong()
        interpolator = FastOutSlowInInterpolator()
        val listener = AnimatorListener()
        addUpdateListener(listener)
        addListener(listener)
    }

    private var nextSubtitle: CharSequence? = null

    constructor(context: Context) : super(context)

    constructor(context: Context, attrs: AttributeSet?) : super(context, attrs)

    constructor(context: Context, attrs: AttributeSet?, @AttrRes defStyleAttr: Int) : super(
        context, attrs, defStyleAttr
    )

    override fun getSubtitle(): CharSequence? = nextSubtitle ?: super.getSubtitle()

    override fun setSubtitle(subtitle: CharSequence?) {
        if (getSubtitle() == subtitle) {
            return
        }
        nextSubtitle = subtitle
        ensureSubtitleAnimatorTarget()
        if (subtitleAnimator.target == null) {
            // Subtitle text view not available (yet), just delegate to super.
            super.setSubtitle(subtitle)
            return
        }
        if (!subtitleAnimator.isRunning) {
            subtitleAnimator.start()
        }
    }

    private fun ensureSubtitleAnimatorTarget() {
        if (subtitleAnimator.target != null) {
            return
        }
        val subtitleTextView = try {
            Toolbar::class.java.getDeclaredField("mSubtitleTextView")
                .apply { isAccessible = true }
                .get(this) as TextView?
        } catch (e: Exception) {
            e.printStackTrace()
            null
        } ?: return
        subtitleAnimator.target = subtitleTextView
    }

    private inner class AnimatorListener : AnimatorListenerAdapter(), AnimatorUpdateListener {
        private var isTextUpdated = false

        override fun onAnimationUpdate(animator: ValueAnimator) {
            if (animator.animatedFraction < 0.5) {
                isTextUpdated = false
            } else {
                ensureTextUpdated()
            }
        }

        override fun onAnimationEnd(animator: Animator) {
            ensureTextUpdated()
            if (nextSubtitle != null) {
                isTextUpdated = false
                animator.start()
            }
        }

        private fun ensureTextUpdated() {
            if (!isTextUpdated) {
                if (nextSubtitle != null) {
                    super@CrossfadeSubtitleToolbar.setSubtitle(nextSubtitle)
                    nextSubtitle = null
                }
                isTextUpdated = true
            }
        }
    }
}
